### PR TITLE
Removed UIntBase to avoid expensive virtual calls for base data type classes

### PR DESCRIPTION
**Pull request by osmirnov**
Calls through virtual tables are quite expensive for such heavily used classes as base data types. Done:
- Removed UIntBase and pushed shared methods implementation to individual class level (UInt160, UInt256)
- Size is immutable now
- Construction from array uses faster Array.Copy method instead LINQ method to ToArray
- Equals methods and comparison operators were optimized by dropping duplicated if checks
- CompareTo uses builtin comparison for arrays instead of looping now
- ToArray returns always a new array to avoid direct access to internal one
- Parse/TryParse methods were delegated to Helper.HexToBytes method
- Unused code that relies on UIntBase and casts it on fly was adjusted
- Tests for UInt160/256 were added (100% coverage)

### CODE REVIEW

File: neo/Wallets/AssetDescriptor.cs [hunk 1]
```diff
@@ -7,13 +7,15 @@ namespace Neo.Wallets
 {
     public class AssetDescriptor
     {
-        public UIntBase AssetId;
+        public object AssetId;
```
Comment (erikzhang) [inline neo/Wallets/AssetDescriptor.cs]: Can we keep `UIntBase` but change everything to non-virtual? Or just change it to an interface?

Comment (osmirnov) [reply neo/Wallets/AssetDescriptor.cs]: We can keep UIntBase or introduce IUInt interface but only shared contract between UInt160 and UInt256 is ToArray method. The rest methods are either type specific or covered by other interfaces (IEquatable, IComparable, ISerializable). I see the consumer of UIntBase/IUInt always has to cast it the underlying type to work with.

For AssetDescriptor class (and for TransferOutput class) we can come up with more descriptive design. The idea here is to move casting logic outside of the class.

Option #1:
```C#
    public class AssetDescriptor<TUInt>
    {
        public TUInt AssetId;
        public string AssetName;
        public byte Decimals;

        public AssetDescriptor(TUInt asset_id)
        {
                AssetId = asset_id
        }
```

Option #2:
```C#
    public class AssetDescriptor256 : IAssetDescriptor
    {
        public UInt256 AssetId;
        public string AssetName;
        public byte Decimals;

        public AssetDescriptor(UInt256 asset_id)
        {
                AssetId = asset_id
        }

        public byte[] IAssetDescriptor.GetAssestIdAsByteArray()
        {
            return AssetId.ToArray();
        }
```

I'd like to elaborate more on the subject but I couldn't have found any example AssetDescriptor class usage.

Comment (erikzhang) [reply neo/Wallets/AssetDescriptor.cs]: The usage of `AssetDescriptor` is in [neo-gui](https://github.com/neo-project/neo-gui) and [neo-cli](https://github.com/neo-project/neo-cli).

**Comment by osmirnov** [comment]
I reviewed neo-gui and neo-cli projects and reverted UIntBase back to API they use. It was discovered to be inheritance from UIntBase and UIntBase.Parse method. Let me know if I missed any.

**Pull request closed without merging**